db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -2210,6 +2239,22 @@ Status DBImpl::FlushMemTable(ColumnFamilyData* cfd,
       if (cfd->imm()->NumNotFlushed() != 0 || !cfd->mem()->IsEmpty() ||
           !cached_recoverable_state_empty_.load()) {
         FlushRequest req{flush_reason, {{cfd, flush_memtable_id}}};
+        if (flush_options.wait &&
+            ShouldRescheduleFlushRequestToRetainUDT(req)) {
+          std::ostringstream oss;
+          oss << "Flush cannot continue until all contained user-defined "
+                 "timestamps are above full_history_ts_low. Increase"
+                 "full_history_ts_low and try again. Or set FlushOptions.wait"
+                 "to false and let background flush jobs automatically"
+                 "continue when it detects full_history_ts_low increase.";
+          if (needs_to_join_write_thread) {
+            write_thread_.ExitUnbatched(&w);
+            if (two_write_queues_) {
+              nonmem_write_thread_.ExitUnbatched(&nonmem_w);
+            }
+          }
+          return Status::TryAgain(oss.str());
+        }
```

Comment:
That's a good point. The write stall condition can vary on the initial manual flush request and the time background check happens. This is intended to make sure the user using the UDT in memtable only feature understand the implications when they are doing a manual flush. To make this check and the background check more consistent, this is updated to only examine whether the keys have expired. If not, it will ask the users to explicitly advance the `full_history_ts_low` if they set wait to true and try again,  since the background flush will automatically advance the cutoff timestamp anyway. What do you think?

Reply:
I think it is better in that it can no longer be impacted by a transient condition (write stalls), whose behavior may differ in testing/dev vs. production. Now I wonder though how the user can use Flush() with wait=true without actually stopping writes. Won't we need to postpone if there is even one write between increasing full_history_ts_low and the point Flush() reaches SwitchMemtable()? It feels like a difficult requirement to actually stop writes, and since they specified wait==true  I'd let them wait

Reply:
That makes sense, in this scenario you described, users will get TryAgain error again, which is very confusing. And we cannot avoid that without making the writes stop. I have removed this check let the manual Flush request continue as usual and leave all the checks to the background work.